x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/PreAnalyzer.java

```diff
@@ -43,17 +44,19 @@ public PreAnalysis preAnalyze(LogicalPlan plan) {
     }
 
     protected PreAnalysis doPreAnalyze(LogicalPlan plan) {
-        Holder<IndexMode> indexMode = new Holder<>();
-        Holder<IndexPattern> indexPattern = new Holder<>();
+        Map<IndexPattern, IndexMode> indexes = new HashMap<>();
         List<IndexPattern> lookupIndices = new ArrayList<>();
         plan.forEachUp(UnresolvedRelation.class, p -> {
             if (p.indexMode() == IndexMode.LOOKUP) {
                 lookupIndices.add(p.indexPattern());
-            } else if (indexMode.get() == null || indexMode.get() == p.indexMode()) {
-                indexMode.set(p.indexMode());
-                indexPattern.set(p.indexPattern());
+            } else if (indexes.containsKey(p.indexPattern()) == false || indexes.get(p.indexPattern()) == p.indexMode()) {
+                indexes.put(p.indexPattern(), p.indexMode());
```

Comment:
Not sure I get why we do the put if we already have exactly the same element already. Can't we simplify this as: 
```
            } else {
                IndexMode previous = indexes.put(p.indexPattern(), p.indexMode());
                if(previous != null && previous != p.indexMode()) {
                    throw new IllegalStateException(
                        "index pattern '" + p.indexPattern() + "' found with with different index mode: " + previous + " != " + p.indexMode()
                    );
                }
            }
```

Also, when exactly can such a situation happen anyway? And if the same pattern can resolve to different modes, why it's OK if one of those modes is lookup, but not OK otherwise?

Reply:
This code was sort-of inherited from the previous version. I tried to keep the logic as close as possible. However, Ievgen also questioned the need for such defensive coding, and suggested I consider an assert instead. I'll investigate that option.